### ISSUE TITLE
Do you really want to be using "with logits"?

### ISSUE THREAD

**Issue by SimonBiggs** [matthewdeancooper/lightning_proto#1]
https://github.com/matthewdeancooper/lightning_proto/blob/649b6bb8ac61daf607a0ef2c8024ac220db6f1b5/lightning_proto/model2D.py#L72

I thought our masks were between 0 and 1 and the last sigmoid activation of the model ensures as such...

**Comment by SimonBiggs** [comment]
Also, if binary cross entropy is the loss function it can accept numbers between 0 and 1 as it's mask, so, a mask need not be binary :).

See https://colab.research.google.com/github/pymedphys/pymedphys/blob/54926066771e6a30bbf03b3a741aa22d3044f316/prototyping/unet.ipynb

**Comment by matthewdeancooper** [comment]
> https://github.com/matthewdeancooper/lightning_proto/blob/649b6bb8ac61daf607a0ef2c8024ac220db6f1b5/lightning_proto/model2D.py#L72
> 
> I thought our masks were between 0 and 1 and the last sigmoid activation of the model ensures as such...

Cheers @SimonBiggs.

From my research the advice is to pass to binary cross entropy in logit form as this improves numerical stability. You're correct if we were to use the dice coefficient as our loss function. However, we can add a sigmoid activation as an initial step in the dice function as this doesn't have any trainable parameters.

**Comment by SimonBiggs** [comment]
> improves numerical stability

I wonder what that means?

**Comment by matthewdeancooper** [comment]
> > improves numerical stability
> 
> I wonder what that means?

I believe it has something to do with calculating the sigmoid on raw inputs to scale between [0, 1] only to then perform an inverse operation (back to raw values). The issue is that the inverse may not span the full range of raw inputs for extreme value cases.

**Comment by SimonBiggs** [comment]
Hmm... I still don't understand. It's okay though. On my rough empirical experimentation I was able to converge to a solution quite a bit quicker when I didn't use the "from_logits" flag in the following case:

https://colab.research.google.com/github/pymedphys/pymedphys/blob/54926066771e6a30bbf03b3a741aa22d3044f316/prototyping/unet.ipynb

Also, including `from_logits` causes some quite weird states. Without `from_logits` tends to be far far more stable...
